### PR TITLE
Use -workDir on gopkgs to improve performance of packages listing

### PR DESCRIPTION
**Pull request by uudashr**
Use `-workDir` this should improve the performance.

The difference on my case:
Without -workDir
```
time gopkgs |wc -l
    5822
gopkgs  1.95s user 2.57s system 121% cpu 3.740 total
wc -l  0.00s user 0.00s system 0% cpu 3.739 total
```

vs
Using -workDir
```
time gopkgs -workDir=.|wc -l
     836
gopkgs -workDir=.  0.28s user 0.31s system 225% cpu 0.261 total
wc -l  0.00s user 0.00s system 1% cpu 0.259 total
```

The duration 3.740 vs 0.261, the different around 3 sec.

### CODE REVIEW

**Comment by uudashr** [comment]
Test on server failed.

I'cant test on my local machine, using node version 10. I always got `TypeError: Cannot read property 'packageJSON' of undefined`.
Not sure it caused by node or by the vscode. @ramya-rao-a maybe you know anything about this?

**Comment by uudashr** [comment]
Nevermind.. fixed. Just need to delete the yarn.lock and node_modules

**Comment by uudashr** [comment]
@ramya-rao-a please check

File: src/goPackages.ts [hunk 1]
```diff
@@ -66,23 +77,27 @@ function gopkgs(): Promise<Map<string, string>> {
...
+		}
+		subs.push(callback);
 
-				let subs = gopkgsSubscriptions;
-				gopkgsSubscriptions = [];
+		// Ensure only single gokpgs running
+		if (!gopkgsRunning.has(workDir)) {
+			gopkgsRunning.add(workDir);
 
+			gopkgs().then((pkgMap) => {
```
Comment (ramya-rao-a) [inline src/goPackages.ts]: shouldn't we pass the `workDir` to `gopkgs` here?

File: src/goPackages.ts [hunk 2]
```diff
@@ -2,21 +2,32 @@ import vscode = require('vscode');
...
+
+let allPkgsCache: Map<string, Cache> = new Map<string, Cache>();
+
+function gopkgs(workDir?: string): Promise<Map<string, string>> {
 	let t0 = Date.now();
 	return new Promise<Map<string, string>>((resolve, reject) => {
-		const cmd = cp.spawn(getBinPath('gopkgs'), ['-format', '{{.Name}};{{.ImportPath}}'], { env: getToolsEnvVars() });
+		const args = ['-format', '{{.Name}};{{.ImportPath}}'];
+		if (workDir) {
+			args.push('-workDir', workDir);
```
Comment (ramya-rao-a) [inline src/goPackages.ts]: This will fail for users who don't have the latest `gopkgs` i.e they have a version where `workDir` is not a supported.
You'll have to something similar to https://github.com/Microsoft/vscode-go/blob/0.6.79/src/goFormat.ts#L33 to prompt the user to update their tool, and fallback to not using the `workDir`

File: src/goPackages.ts [hunk 3]
```diff
@@ -91,13 +106,15 @@ function getAllPackagesNoCache(): Promise<Map<string, string>> {
...
+	let cache = allPkgsCache.get(workDir);
+	let useCache = cache && (new Date().getTime() - cache.lastHit) < cacheTimeout;
 	if (useCache) {
-		allPkgsLastHit = new Date().getTime();
-		return Promise.resolve(allPkgsCache);
+		cache.lastHit = new Date().getTime();
+		return Promise.resolve(cache.entry);
 	}
 
 	return getAllPackagesNoCache().then((pkgs) => {
```
Comment (ramya-rao-a) [inline src/goPackages.ts]: shouldn't we pass the `workDir` here?

File: src/goPackages.ts (hunk 2 above)
Comment (ramya-rao-a) [inline src/goPackages.ts]: This will fail for users who don't have the latest `gopkgs` i.e they have a version where `workDir` is not a supported.
You'll have to something similar to https://github.com/Microsoft/vscode-go/blob/0.6.79/src/goFormat.ts#L33 to prompt the user to update their tool, and fallback to not using the `workDir`

File: src/goPackages.ts (hunk 1 above)
Comment (ramya-rao-a) [inline src/goPackages.ts]: pass `workDir` here

File: src/goPackages.ts (hunk 3 above)
Comment (ramya-rao-a) [inline src/goPackages.ts]: pass `workDir` here

**Comment by uudashr** [comment]
Wow... I missed a lot. Fixed.

File: src/goPackages.ts (hunk 2 above)
Comment (uudashr) [reply src/goPackages.ts]: Done

File: src/goPackages.ts (hunk 1 above)
Comment (uudashr) [reply src/goPackages.ts]: done

File: src/goPackages.ts (hunk 3 above)
Comment (uudashr) [reply src/goPackages.ts]: done

File: src/goImport.ts [hunk 4]
```diff
@@ -13,9 +13,9 @@ import { getImportablePackages } from './goPackages';
 
 const missingToolMsg = 'Missing tool: ';
 
-export function listPackages(excludeImportedPkgs: boolean = false): Thenable<string[]> {
+export function listPackages(excludeImportedPkgs: boolean = false, workDir?: string): Thenable<string[]> {
```
Comment (ramya-rao-a) [inline src/goImport.ts]: why is this change needed when `getImportablePackages` can calculate the `workDir` by using the given `fileName`?

File: src/goPackages.ts [hunk 5]
```diff
@@ -66,23 +82,27 @@ function gopkgs(): Promise<Map<string, string>> {
...
+		subs.push(callback);
 
-				let subs = gopkgsSubscriptions;
-				gopkgsSubscriptions = [];
+		// Ensure only single gokpgs running
+		if (!gopkgsRunning.has(workDir)) {
+			gopkgsRunning.add(workDir);
 
+			gopkgs(workDir).then((pkgMap) => {
+				gopkgsRunning.delete(workDir);
```
Comment (ramya-rao-a) [inline src/goPackages.ts]: Entries never get deleted from `gopkgsRunning` when `gopkgs` returns a rejected promise

File: src/goPackages.ts [hunk 6]
```diff
@@ -29,6 +40,11 @@ function gopkgs(): Promise<Map<string, string>> {
 				return promptForMissingTool('gopkgs');
 			}
 
+			if (errchunks && errchunks.join('').startsWith('flag provided but not defined: -workDir')) {
+				promptForUpdatingTool('gopkgs');
+				return Promise.resolve(pkgs);
```
Comment (ramya-rao-a) [inline src/goPackages.ts]: fallback to not using `workdir` by `return gopkgs().then(result => resolve(result))` instead.

File: src/goPackages.ts (hunk 5 above)
Comment (uudashr) [reply src/goPackages.ts]: No rejection on gopkgs. Positive and negative result will be return by `resolve()`.

Base on your comment, negative should be in form of rejection. Such as:
1. When `gopkgs` binary not found
2. When `gopkgs` needs to update

This is part of quality I think, which I'm okay, we can do this. But maybe after this PR.

File: src/goPackages.ts (hunk 6 above)
Comment (uudashr) [reply src/goPackages.ts]: Ahh... good idea

File: src/goImport.ts (hunk 4 above)
Comment (uudashr) [reply src/goImport.ts]: I need this to pass the test of **"Replace vendor packages with relative path"** it will fail on the assert of *"Relative path for vendor package 9fans.net/go/acme not found"*.

This happen because `gopkgs` recognize vscode workspace as `workDir`, and the workspace dir for the running test is `vscode-go/out/test`, no vendor package under the directory. While actually we are testing vendor package on under `$GOPATH/src/github.com/rogpeppe/godef`, there the vendor package of `9fans.net/go/acme` should be found.

So it only for satisfying the test. The actual running vscode will never use `workDir` param or `listPackages()` (which kind of silly). I can't guess the project root/`workDir` based on  `fileName`. Any suggestion?

Available the options are:
1. Guess golang project dir by filename
2. Change the test workspace dir to `$GOPATH/src/github.com/rogpeppe/godef` above test
3. Pass `workDir` to `listPackages()` (currently use this)

File: src/goPackages.ts (hunk 6 above)
Comment (uudashr) [reply src/goPackages.ts]: Done

File: src/goImport.ts (hunk 4 above)
Comment (ramya-rao-a) [reply src/goImport.ts]: @uudashr This can happen IRL too. A user might open and edit a Go file that is not under the current workspace.

File: src/goPackages.ts [hunk 7]
```diff
@@ -66,23 +83,27 @@ function gopkgs(): Promise<Map<string, string>> {
...
+		if (!subs) {
+			subs = [];
+			gopkgsSubscriptions.set(workDir, subs);
+		}
+		subs.push(callback);
 
-				let subs = gopkgsSubscriptions;
-				gopkgsSubscriptions = [];
+		// Ensure only single gokpgs running
+		if (!gopkgsRunning.has(workDir)) {
```
[the remark this answers is not in the document]
Comment (uudashr) [reply src/goPackages.ts]: Since it already run, we can use the cache.

**Comment by uudashr** [comment]
Fail on test. Mac machine on Travis CI seems out of resources

**Comment by uudashr** [comment]
All good now

File: src/goPackages.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (uudashr) [reply src/goPackages.ts]: Made some changes... all the go files under the same package root will share the same cache and `gopkgs` running  instance.

[the remark this answers is not in the document]
Comment (ramya-rao-a) [reply src/goPackages.ts]: @uudashr I was **just** thinking about the same thing over the weekend :) The fact that vendor packages under a higher ancestor is also available to the current file

[the remark this answers is not in the document]
Comment (uudashr) [reply src/goPackages.ts]: Ahaha.. cool 👍

**Comment by uudashr** [comment]
@ramya-rao-a which test? on travis all good.
Is there a way to reproduce?

**Comment by uudashr** [comment]
Issues on gopkgs... for symbolic link

**Comment by uudashr** [comment]
Issues [godirwalk](https://github.com/karrick/godirwalk/pull/4). Just send the owner a PR https://github.com/karrick/godirwalk/pull/4.

Meanwhile the `gopkgs` use the patched version and already work as expected.

@ramya-rao-a could you re-run the test on your machine?

**Comment by ramya-rao-a** [comment]
Re-installing `gopkgs` fails with "unexpected directory layout" for the `godirwalk`. Had to delete the `gopkgs` folder and then install from scratch

**Comment by uudashr** [comment]
Due to this issue https://github.com/golang/go/issues/21057
You have to execute another `go get -u -github.com/uudashr/gopkgs/cmd/gopkgs` if the first attempt fail. 

Try to add -v `go get -u -v -github.com/uudashr/gopkgs/cmd/gopkgs` so we know what happen.

Let me know if still fail, I have to find another way then.

**Comment by ramya-rao-a** [comment]
Second update worked. I'll run some tests tomorrow and get back here.

File: src/goPackages.ts [hunk 8]
```diff
@@ -149,12 +182,21 @@ export function getImportablePackages(filePath: string, useCache: boolean = fals
 				return;
 			}
 
-			let relativePkgPath = getRelativePackagePath(currentFileDirPath, currentWorkspace, pkgPath);
+			if (!foundPkgRootDir) {
```
Comment (ramya-rao-a) [inline src/goPackages.ts]: @uudashr 
You will have to find the `pkgRootDir` upfront and not after getting the results from `gopkgs`. To understand why, take the below case

- If you dont have the `godef` code, then run `go get github.com/rogpeppe/godef` 
- Open the `godef` folder and then open the file `ast.go` in it
- Run the command `Go: Add Import` with the current Go extension and then with your changes and search for `acme`
- In the current Go extension, you will get the result for "9fans.net/go/acme", but you wont get the same with your changes.

This is because, the `acme` package is in the vendor folder which is not a sibling of the directory to which the file `ast.go` file belongs.

Comment (uudashr) [reply src/goPackages.ts]: No, it should works.

Don't know which version `gopkgs` did you use. Since I push the latest changes, gopkgs `-workDir` also works on sub packages. Try to update the `gopkgs`

```
$ gopkgs -format '{{.Name}};{{.ImportPath}}' -workDir=$(go env GOPATH)/src/github.com/rogpeppe/godef/go/ast|grep acme
acme;github.com/rogpeppe/godef/vendor/9fans.net/go/acme
```

And this also covered on the test case *"Replace vendor packages with relative path"*

Comment (ramya-rao-a) [reply src/goPackages.ts]: Works in my Mac but not Windows

Comment (uudashr) [reply src/goPackages.ts]: Just push an update... would you mind to test it on Windows?

**Comment by ramya-rao-a** [comment]
@uudashr That's because linting was failing on master which I just fixed. Sorry about that

**Comment by uudashr** [comment]
@ramya-rao-a no problem. If you have time please run the test on Windows.

**Comment by ramya-rao-a** [comment]
@uudashr The problem in Windows needed your fix as well as https://github.com/Microsoft/vscode-go/pull/1658/commits/0055c20a80922c1261bf64398a1c1fbea6395240.

I made one more commit to print the error message from gopkgs to the console. This should help in troubleshooting I ran into `gopkgs` failure myself a few times as discussed in this PR.

I also added a retry logic for when the "unexpected directory layout" error is encountered.

The unit tests are passing on my Windows machine.

**Comment by ramya-rao-a** [comment]
I've updated https://github.com/Microsoft/vscode-go/issues/1490 with the means to try out these changes
